### DISCUSSION TITLE
Bit-Perfect Playback

### DISCUSSION THREAD

**Discussion by andrewdunndev** [jeffvli/feishin#88] [General]
Wanted to post here about bit-perfect playback since it was _slightly_ off the beaten path.

## Linux 

Going to use alsa as playback subsystem, does mean you need an install of mpv that supports alsa. For example the flathub install by default does not, uses pipewire instead.

What sound devices do we have: 
```
[agd@small ~]$ cat /proc/asound/cards 
 0 [Generic_1      ]: HDA-Intel - HD-Audio Generic
                      HD-Audio Generic at 0xfcec8000 irq 82
 1 [Generic        ]: HDA-Intel - HD-Audio Generic
                      HD-Audio Generic at 0xfcec0000 irq 83
 2 [C925e          ]: USB-Audio - Logitech Webcam C925e
                      Logitech Webcam C925e at usb-0000:01:00.0-1, high speed
 3 [Multibit       ]: USB-Audio - Schiit USB Multibit
                      Schiit Audio Schiit USB Multibit at usb-0000:01:00.0-2, high speed
 4 [Device         ]: USB-Audio - USB Audio Device
                      C-Media Electronics Inc. USB Audio Device at usb-0000:01:00.0-6, full speed
```

We're going to use the __Multibit__ device:

```
[agd@small ~]$ mpv --audio-device=help | grep Multibit
  'pipewire/alsa_output.usb-Schiit_Audio_Schiit_USB_Multibit-00.iec958-stereo' (CM6631A Audio Processor Digital Stereo (IEC958))
  'pulse/alsa_output.usb-Schiit_Audio_Schiit_USB_Multibit-00.iec958-stereo' (CM6631A Audio Processor Digital Stereo (IEC958))
  'alsa/sysdefault:CARD=Multibit' (Schiit USB Multibit, USB Audio/Default Audio Device)
  'alsa/front:CARD=Multibit,DEV=0' (Schiit USB Multibit, USB Audio/Front output / input)
  'alsa/surround21:CARD=Multibit,DEV=0' (Schiit USB Multibit, USB Audio/2.1 Surround output to Front and Subwoofer speakers)
  'alsa/surround40:CARD=Multibit,DEV=0' (Schiit USB Multibit, USB Audio/4.0 Surround output to Front and Rear speakers)
  'alsa/surround41:CARD=Multibit,DEV=0' (Schiit USB Multibit, USB Audio/4.1 Surround output to Front, Rear and Subwoofer speakers)
  'alsa/surround50:CARD=Multibit,DEV=0' (Schiit USB Multibit, USB Audio/5.0 Surround output to Front, Center and Rear speakers)
  'alsa/surround51:CARD=Multibit,DEV=0' (Schiit USB Multibit, USB Audio/5.1 Surround output to Front, Center, Rear and Subwoofer speakers)
  'alsa/surround71:CARD=Multibit,DEV=0' (Schiit USB Multibit, USB Audio/7.1 Surround output to Front, Center, Side, Rear and Woofer speakers)
  'alsa/iec958:CARD=Multibit,DEV=0' (Schiit USB Multibit, USB Audio/IEC958 (S/PDIF) Digital Audio Output)
```

We'll test with mpv directly, using a 24/96 flac file:

```
mpv --audio-device=alsa/iec958:CARD=Multibit,DEV=0 --audio-exclusive=yes --gapless-audio=no  New\ Love.flac
```
verify that the hardware is indeed switching sample rates:
```
[agd@small ~]$ cat /proc/asound/card3/pcm0p/sub0/hw_params 
access: RW_INTERLEAVED
format: S32_LE
subformat: STD
channels: 2
rate: 96000 (96000/1)
period_size: 2400
buffer_size: 9600
```

Now with __feishin__ we have to set some parameters:

 - MPV parameters: `--audio-device=alsa/iec958:CARD=Multibit,DEV=0`
 - Gapless audio: no
 - Audio Exclusive Mode: toggled on

![image](https://user-images.githubusercontent.com/323695/233469123-04113c25-2368-4b8f-9ab9-d88d6b7723e7.png)

## Apple

Not verified yet because I don't know how to query the audio hardware to see if its sample rate switching. The DACs I have right now don't have physical indicators on them. Need to be aware of [coreaudio](https://mpv.io/manual/stable/#audio-output-drivers-coreaudio) options and [audio output](https://mpv.io/manual/stable/#audio-output-drivers). Potentially just need these two parameters added:

`--ao=coreaudio_exclusive --coreaudio-change-physical-format=yes`

```
agd@mini ~ % system_profiler SPAudioDataType
Audio:

    Devices:

        LG TV SSCR2:

          Manufacturer: GSM
          Output Channels: 2
          Current SampleRate: 48000
          Transport: HDMI
          Output Source: Default

        Schiit Unison Modius:

          Manufacturer: Schiit Audio
          Output Channels: 2
          Current SampleRate: 192000
          Transport: USB
          Output Source: Default

        Mac mini Speakers:

          Default Output Device: Yes
          Default System Output Device: Yes
          Manufacturer: Apple Inc.
          Output Channels: 2
          Current SampleRate: 48000
          Transport: Built-in
          Output Source: Mac mini Speakers
```

**Comment by Aaronlonin** [comment]
Is this supported on Mac? I'm trying Feishin for the first time today and noticed that when selecting Web as Audio Player, there's no exclusive mode. But when switching to MPV, the Audio Device dropdown becomes empty (it does not detect any of my devices). I tried manually setting the audio device by passing it as a parameter with `--audio-device="FiiO K11"`, but it doesn't seem to pick it up. Any recomendation?

> Wanted to post here about bit-perfect playback since it was _slightly_ off the beaten path.
> 
> ## Linux
> Going to use alsa as playback subsystem, does mean you need an install of mpv that supports alsa. For example the flathub install by default does not, uses pipewire instead.
> 
> What sound devices do we have:
> 
> ```
> [agd@small ~]$ cat /proc/asound/cards 
>  0 [Generic_1      ]: HDA-Intel - HD-Audio Generic
>                       HD-Audio Generic at 0xfcec8000 irq 82
>  1 [Generic        ]: HDA-Intel - HD-Audio Generic
>                       HD-Audio Generic at 0xfcec0000 irq 83
>  2 [C925e          ]: USB-Audio - Logitech Webcam C925e
>                       Logitech Webcam C925e at usb-0000:01:00.0-1, high speed
>  3 [Multibit       ]: USB-Audio - Schiit USB Multibit
>                       Schiit Audio Schiit USB Multibit at usb-0000:01:00.0-2, high speed
>  4 [Device         ]: USB-Audio - USB Audio Device
>                       C-Media Electronics Inc. USB Audio Device at usb-0000:01:00.0-6, full speed
> ```
> 
> We're going to use the **Multibit** device:
> 
> ```
> [agd@small ~]$ mpv --audio-device=help | grep Multibit
>   'pipewire/alsa_output.usb-Schiit_Audio_Schiit_USB_Multibit-00.iec958-stereo' (CM6631A Audio Processor Digital Stereo (IEC958))
>   'pulse/alsa_output.usb-Schiit_Audio_Schiit_USB_Multibit-00.iec958-stereo' (CM6631A Audio Processor Digital Stereo (IEC958))
>   'alsa/sysdefault:CARD=Multibit' (Schiit USB Multibit, USB Audio/Default Audio Device)
>   'alsa/front:CARD=Multibit,DEV=0' (Schiit USB Multibit, USB Audio/Front output / input)
>   'alsa/surround21:CARD=Multibit,DEV=0' (Schiit USB Multibit, USB Audio/2.1 Surround output to Front and Subwoofer speakers)
>   'alsa/surround40:CARD=Multibit,DEV=0' (Schiit USB Multibit, USB Audio/4.0 Surround output to Front and Rear speakers)
>   'alsa/surround41:CARD=Multibit,DEV=0' (Schiit USB Multibit, USB Audio/4.1 Surround output to Front, Rear and Subwoofer speakers)
>   'alsa/surround50:CARD=Multibit,DEV=0' (Schiit USB Multibit, USB Audio/5.0 Surround output to Front, Center and Rear speakers)
>   'alsa/surround51:CARD=Multibit,DEV=0' (Schiit USB Multibit, USB Audio/5.1 Surround output to Front, Center, Rear and Subwoofer speakers)
>   'alsa/surround71:CARD=Multibit,DEV=0' (Schiit USB Multibit, USB Audio/7.1 Surround output to Front, Center, Side, Rear and Woofer speakers)
>   'alsa/iec958:CARD=Multibit,DEV=0' (Schiit USB Multibit, USB Audio/IEC958 (S/PDIF) Digital Audio Output)
> ```
> 
> We'll test with mpv directly, using a 24/96 flac file:
> 
> ```
> mpv --audio-device=alsa/iec958:CARD=Multibit,DEV=0 --audio-exclusive=yes --gapless-audio=no  New\ Love.flac
> ```
> 
> verify that the hardware is indeed switching sample rates:
> 
> ```
> [agd@small ~]$ cat /proc/asound/card3/pcm0p/sub0/hw_params 
> access: RW_INTERLEAVED
> format: S32_LE
> subformat: STD
> channels: 2
> rate: 96000 (96000/1)
> period_size: 2400
> buffer_size: 9600
> ```
> 
> Now with **feishin** we have to set some parameters:
> 
> * MPV parameters: `--audio-device=alsa/iec958:CARD=Multibit,DEV=0`
> * Gapless audio: no
> * Audio Exclusive Mode: toggled on
> 
> ![image](https://user-images.githubusercontent.com/323695/233469123-04113c25-2368-4b8f-9ab9-d88d6b7723e7.png)
> 
> ## Apple
> Not verified yet because I don't know how to query the audio hardware to see if its sample rate switching. The DACs I have right now don't have physical indicators on them. Need to be aware of [coreaudio](https://mpv.io/manual/stable/#audio-output-drivers-coreaudio) options and [audio output](https://mpv.io/manual/stable/#audio-output-drivers). Potentially just need these two parameters added:
> 
> `--ao=coreaudio_exclusive --coreaudio-change-physical-format=yes`
> 
> ```
> agd@mini ~ % system_profiler SPAudioDataType
> Audio:
> 
>     Devices:
> 
>         LG TV SSCR2:
> 
>           Manufacturer: GSM
>           Output Channels: 2
>           Current SampleRate: 48000
>           Transport: HDMI
>           Output Source: Default
> 
>         Schiit Unison Modius:
> 
>           Manufacturer: Schiit Audio
>           Output Channels: 2
>           Current SampleRate: 192000
>           Transport: USB
>           Output Source: Default
> 
>         Mac mini Speakers:
> 
>           Default Output Device: Yes
>           Default System Output Device: Yes
>           Manufacturer: Apple Inc.
>           Output Channels: 2
>           Current SampleRate: 48000
>           Transport: Built-in
>           Output Source: Mac mini Speakers
> ```

**Reply by thearrow** [reply]
> Is this supported on Mac? I'm trying Feishin for the first time today and noticed that when selecting Web as Audio Player, there's no exclusive mode.

@Aaronlonin Here's how I'm doing this on a Mac - I leave all the Feishin GUI options untouched and only specify the MPV args. My DAC is configured as the default audio device on my Mac (at the OS level) and it just works 🤷  

My DAC (also a FiiO) has an LCD display that indicates the sample rate, and MPV successfully adjusts it to match what is playing. This also prevents any other audio coming from other apps (the only way I know how to verify exclusive mode).

<img width="1878" height="1550" alt="CleanShot 2026-02-17 at 16 28 12@2x" src="https://github.com/user-attachments/assets/c034dad6-f037-46bb-8fe6-661ba2d102bf" />

Hope that helps!

**Reply by Aaronlonin** [reply]
Dude youre fast!! I really appreciate you responding to my question. So I got help from some IA, and it turns out that Feishin couldn't find the MPV binary by default, so I manually added it to the config.json file, and it worked like a charm (the dropdown now lists all my devices). And I also confirmed that now it indicates the right sample rates as well, depending on the track.

`"mpv_path": "/opt/homebrew/bin/mpv"`

I could give your list of parameters a try too if you think that could help others (for the sake of science).

> > Is this supported on Mac? I'm trying Feishin for the first time today and noticed that when selecting Web as Audio Player, there's no exclusive mode.
> 
> @Aaronlonin Here's how I'm doing this on a Mac - I leave all the Feishin GUI options untouched and only specify the MPV args. My DAC is configured as the default audio device on my Mac (at the OS level) and it just works 🤷
> 
> My DAC (also a FiiO) has an LCD display that indicates the sample rate, and MPV successfully adjusts it to match what is playing. This also prevents any other audio coming from other apps (the only way I know how to verify exclusive mode).
> 
> <img alt="CleanShot 2026-02-17 at 16 28 12@2x" width="1878" height="1550" src="https://private-user-images.githubusercontent.com/447736/551198180-c034dad6-f037-46bb-8fe6-661ba2d102bf.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzEzNjUwODgsIm5iZiI6MTc3MTM2NDc4OCwicGF0aCI6Ii80NDc3MzYvNTUxMTk4MTgwLWMwMzRkYWQ2LWYwMzctNDZiYi04ZmU2LTY2MWJhMmQxMDJiZi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjYwMjE3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI2MDIxN1QyMTQ2MjhaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01MTJlOWJlMzlhY2VlMTlkYjIyNjFiZTIzZGQ2MDJmNTVjOTU1NDk2NmI5YjM0MjU5Y2UxN2MzMjEyMTJmNWVhJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.0Rj888HlYGbTdWC4XowTQfBfrpZyi56-TubovrSa26U">
> Hope that helps!

**Reply by Aaronlonin** [reply]
For the record, I just tried the followint parameters
```
--ao=coreaudio_exclusive
--audio-exclusive=yes
--coreaudio-change-physical-format=yes
```

And it behaves unexpectedly: Some songs are played in my DAC and some in my monitor (connected by HDMI). It might be something else, but when I remove them, it just works again.

**Reply by thearrow** [reply]
Weird. Oh well, glad you got it figured out with the MPV path!
